torch/_dynamo/eval_frame.py

```diff
@@ -101,20 +101,12 @@ def _maybe_init_guarded_backend_cache():
         guarded_backend_cache.skip_backend_check_for_run_only_mode = False
     if not hasattr(guarded_backend_cache, "current_backend"):
         guarded_backend_cache.current_backend = None
-    if not hasattr(guarded_backend_cache, "cached_backends"):
-        guarded_backend_cache.cached_backends = {}
 
 
 def _reset_guarded_backend_cache():
     _maybe_init_guarded_backend_cache()
     guarded_backend_cache.skip_backend_check_for_run_only_mode = False
     guarded_backend_cache.current_backend = None
-    cached_backends = guarded_backend_cache.cached_backends
-    for backend in cached_backends.values():
-        if hasattr(backend, "reset"):
-            backend.reset()
```

Comment:
oh, maybe this is a problem? We apparently need to call `backend.reset`.

Reply:
But basically, this makes `torch._dynamo.reset` inherently _not thread safe_

Reply:
I think the backend cache needs to be in a global dict, maybe behind a RWLock in that case.

Reply:
@ydwu4 I don't actually see any backend use this `reset` API. Could you comment on this?

Because we mix a lot of global and thread-local state, it is actually completely unclear what semantics `torch._dynamo.reset` has in a multithreaded setting. For instance, if another thread is compiling a function and you call reset, what does that do?

Reply:
Probably the most correct way to handle this is to have a mutex around reset, and then you can have a global dict if you want.

Reply:
But if reset is not actually used by backends, IMO this PR has a better compartmentalization of responsiblities.

Reply:
>  I don't actually see any backend use this reset API.

Inductor backend has logic to reset the cudagraphs [here](https://github.com/pytorch/pytorch/blob/main/torch/__init__.py#L1663).

> Probably the most correct way to handle this is to have a RWLock around reset

Yeah, I think that's what we should do. I discussed with Jason and he shared the same opinion. dynamo has an existing RLock. We can surround reset region with RLock as as a starting point. We could switch to RWLock as a follow up if necessary.

Reply:
Let's make torch._dynamo.reset use the same lock as dynamo compile.